### PR TITLE
Update guidelines for NullableResponse<T>

### PR DESCRIPTION
**Pull request by christothes**

### CODE REVIEW

File: docs/dotnet/introduction.md [hunk 1]
```diff
@@ -439,7 +439,11 @@ Virtual methods are used to support mocking. See [Support for Mocking](#dotnet-m
 
 ##### Return Types
 
-{% include requirement/MUST id="dotnet-service-methods-response-sync" %} return `Response<T>` or `Response` from synchronous methods.
+{% include requirement/MUST id="dotnet-service-methods-response-sync" %} return `Response<T>`, `NullableResponse<T>`, or `Response` from synchronous methods.
```
Comment (JoshLove-msft) [inline docs/dotnet/introduction.md:442]: Is there an async section we need to update?

Comment (christothes) [reply docs/dotnet/introduction.md:442]: The discussion of async is just below and is agnostic to the response type, I believe.

Comment (JoshLove-msft) [reply docs/dotnet/introduction.md:442]: Doesn't look like it is agnostic - https://github.com/Azure/azure-sdk/pull/5028/files#diff-d1668aaa12d853c4d2a8d1bf803705a97fcbf6d557f2be15d17623030a556b7dR450

Comment (christothes) [reply docs/dotnet/introduction.md:442]: OK, yeah maybe I should add the additional type there and move this section below that.

File: docs/dotnet/introduction.md [hunk 2]
```diff
@@ -439,11 +439,15 @@ Virtual methods are used to support mocking. See [Support for Mocking](#dotnet-m
...
 
 `T` represents the content of the response, as described below.
 
-{% include requirement/MUST id="dotnet-service-methods-response-async" %} return `Task<Response<T>>` or `Task<Response>` from asynchronous methods that make network requests.
+{% include requirement/MUST id="dotnet-service-methods-response-async" %} return `Task<Response<T>>`, `Task<NullableResponse<T>>` or `Task<Response>` from asynchronous methods that make network requests.
+
+NOTE: `NullableResponse<T>` and `Task<NullableResponse<T>>` are intended for scenarios where the Response may or may not contain a value. Common examples include:
```
Comment (annelo-msft) [inline docs/dotnet/introduction.md:448]:
```suggestion
{% include note.html content="`NullableResponse<T>` and `Task<NullableResponse<T>>` are intended for scenarios where the Response may or may not contain a value. Common examples include:
```

File: docs/dotnet/introduction.md [hunk 3]
```diff
@@ -439,11 +439,15 @@ Virtual methods are used to support mocking. See [Support for Mocking](#dotnet-m
...
+- Get*IfExists methods in which the value will be returned only if it exists
+- Conditional response APIs such as Get APIs which return no content if the requested resource's `ETag` matches the `If-Match` header
```
Comment (annelo-msft) [inline docs/dotnet/introduction.md:450]:
```suggestion
- Conditional response APIs such as Get APIs which return no content if the requested resource's `ETag` matches the `If-Match` header" %}
```

File: docs/dotnet/introduction.md (hunk 2 above)
Comment (annelo-msft) [inline docs/dotnet/introduction.md:448]: Maybe consider explaining the semantics of `Response<T>` here, to explain `NullableResponse<T>` in contrast?  For example, maybe something like - NOTE: Returning `Response<T>` ensures the caller that `response.Value` will not be null.  `NullableResponse<T>` indicates that `response.HasValue` must be checked before accessing `response.Value`.  `NullableResponse<T>` should be used in scenarios such as ..."

Comment (annelo-msft) [inline docs/dotnet/introduction.md:448]: I'm curious whether we want to be more prescriptive about when and when not to use NullableResponse<T>.  We've given some examples where a library author might use it, but nothing saying they SHOULD or MUST in these cases.  Do we want to enforce consistency on this, or are we ok if people use it sometimes and sometimes not?  And if we do want to enforce consistency, if we could write static analyzers (maybe one for a method suffixed with "IfExists" requiring to return NullableResponse<T>, e.g.), that would help automate that.

File: docs/dotnet/introduction.md (hunk 3 above)
Comment (annelo-msft) [inline docs/dotnet/introduction.md:450]: Would it make sense to call out NullableResponse<T> in the section on Conditional Requests? https://azure.github.io/azure-sdk/dotnet_introduction.html#conditional-request-methods

E.g. DO return NullableResponse<T> when the service could return a 304 response.

?

File: docs/dotnet/introduction.md [hunk 4]
```diff
@@ -439,11 +439,15 @@ Virtual methods are used to support mocking. See [Support for Mocking](#dotnet-m
...
-{% include requirement/MUST id="dotnet-service-methods-response-sync" %} return `Response<T>` or `Response` from synchronous methods.
+{% include requirement/MUST id="dotnet-service-methods-response-sync" %} return `Response<T>`, `NullableResponse<T>`, or `Response` from synchronous methods.
 
 `T` represents the content of the response, as described below.
 
-{% include requirement/MUST id="dotnet-service-methods-response-async" %} return `Task<Response<T>>` or `Task<Response>` from asynchronous methods that make network requests.
+{% include requirement/MUST id="dotnet-service-methods-response-async" %} return `Task<Response<T>>`, `Task<NullableResponse<T>>` or `Task<Response>` from asynchronous methods that make network requests.
+
+NOTE: `NullableResponse<T>` and `Task<NullableResponse<T>>` are intended for scenarios where the Response may or may not contain a value. Common examples include:
+- Get*IfExists methods in which the value will be returned only if it exists
```
Comment (annelo-msft) [inline docs/dotnet/introduction.md:449]: Is `*IfExists` a sufficiently standard naming convention to add it to the list of common method names?
https://azure.github.io/azure-sdk/dotnet_introduction.html#naming

If so, this bullet might be able to link to here.

File: docs/dotnet/introduction.md (hunk 3 above)
Comment (annelo-msft) [reply docs/dotnet/introduction.md:450]: Would we recommend other languages use this convention?

https://azure.github.io/azure-sdk/general_design.html#conditional-requests

We have this guidline:
[⚠️](https://azure.github.io/azure-sdk/general_design.html#) YOU SHOULD NOT throw an error when a 304 Not Modified response is received from the service, unless such errors are idiomatic to the language.

Would it make sense to add that the service method should be able to return a response that doesn't have a body/deserialized value associated?

File: docs/dotnet/introduction.md (hunk 4 above)
Comment (christothes) [reply docs/dotnet/introduction.md:449]: added

File: docs/dotnet/introduction.md (hunk 2 above)
Comment (christothes) [reply docs/dotnet/introduction.md]: added

**Review by annelo-msft — COMMENTED** [review]
Per our side conversation, should we update the guidance to indicate any convenience method that accepts RequestContext should also return NullableResponse\<T\>?

File: docs/dotnet/introduction.md [hunk 5]
```diff
@@ -439,11 +439,15 @@ Virtual methods are used to support mocking. See [Support for Mocking](#dotnet-m
...
 
-{% include requirement/MUST id="dotnet-service-methods-response-sync" %} return `Response<T>` or `Response` from synchronous methods.
+{% include requirement/MUST id="dotnet-service-methods-response-sync" %} return `Response<T>`, `NullableResponse<T>`, or `Response` from synchronous methods.
 
 `T` represents the content of the response, as described below.
 
-{% include requirement/MUST id="dotnet-service-methods-response-async" %} return `Task<Response<T>>` or `Task<Response>` from asynchronous methods that make network requests.
+{% include requirement/MUST id="dotnet-service-methods-response-async" %} return `Task<Response<T>>`, `Task<NullableResponse<T>>` or `Task<Response>` from asynchronous methods that make network requests.
+
+{% include note.html content="`Response<T>` and `Task<Response<T>>` are intended for scenarios in which the response is guaranteed to have a `Value`. In contrast, `NullableResponse<T>` and `Task<NullableResponse<T>>` are intended for scenarios in which the `Response` may or may not contain a value. Common examples include:
```
Comment (KrzysztofCwalina) [inline docs/dotnet/introduction.md:448]: I wonder if we should not simplify this by omitting the Task variants. Task would be implied for async methods.

File: docs/dotnet/introduction.md [hunk 6]
```diff
@@ -665,13 +669,15 @@ public class CopyFromUriOperation {
...
+Some services support conditional requests that are used to implement optimistic concurrency control. In Azure, optimistic concurrency is typically implemented using If-Match headers and ETags. See [Managing Concurrency in Blob Storage](https://docs.microsoft.com/en-us/azure/storage/blobs/concurrency-manage?tabs=dotnet) as a good example.
 
 {% include requirement/MUST id="dotnet-conditional-etag" %} use Azure.Core ETag to represent ETags.
 
 {% include requirement/MAY id="dotnet-conditional-matchcondition" %} take [MatchConditions](https://docs.microsoft.com/en-us/dotnet/api/azure.matchconditions?view=azure-dotnet), [RequestConditions](https://docs.microsoft.com/en-us/dotnet/api/azure.requestconditions?view=azure-dotnet), (or a custom subclass) as a parameter to conditional service call methods.
 
-TODO: more guidelines comming. see https://github.com/Azure/azure-sdk/issues/2154
+{% include requirement/MUST id="dotnet-conditional-nullableresponse" %} return ```NullableResponse<T>``` when the service could return a 304 response with no content.
```
Comment (KrzysztofCwalina) [inline docs/dotnet/introduction.md:678]: ... for example, we don't talk about tasks here.

File: docs/dotnet/introduction.md (hunk 5 above)
Comment (christothes) [reply docs/dotnet/introduction.md]: Done. I just added a bit to the end of the first item to mention Task for async methods.

**Pull request closed without merging**